rj_gameplay/stp/play/__init__.py

```diff
@@ -81,47 +84,67 @@ def assign_roles(
+        Satisfy constraint that all Roles of a Tactic must all be assigned at once.
+        If a Tactic's Roles cannot all be filled, none will be filled and a debug message signifying error is displayed along with the tactic's tick not running."
         """
 
         used_robots = set()
         for tactic in self.prioritized_tactics:
-            # TODO: handle if tactic requests more roles than exist
-            #       by not assigning tactic at all!
-            #       and give some default behavior (later)
+            # Will temporarily hold tactic roles. If any roles cannot be filled, the list will be emptied and an error debug message appears
             robots_for_tactic = []
             for cost_fn, role in tactic.role_requests:
                 min_cost = 1e9
                 cheapest_robot = None
+                # Used in calculating available robots
+                numOfInvisibleRobots = 0
 
+                # Checking to see if any robot can fullfill a specific role request in a specific tactic
                 for robot in world_state.our_robots:
-                    if robot in used_robots or not robot.visible:
+                    if robot in used_robots or robot in robots_for_tactic:
+                        continue
+                    if not robot.visible:
+                        numOfInvisibleRobots += 1
                         continue
                     cost = cost_fn(robot, world_state)
                     if cost < min_cost:
                         min_cost = cost
                         cheapest_robot = robot
 
+                # Occurs if all other visible robots already have a role
                 if cheapest_robot is None:
-                    # TODO: consider looping until success?
-                    raise RoleAssignFailure(
-                        f"RoleRequest ({role}, {cost_fn}) was not assigned"
-                    )
-
-                used_robots.add(cheapest_robot)
+                    print(f"RoleRequest ({role}, {cost_fn}) was not assigned")
+                    robots_for_tactic = None
+                    break
                 robots_for_tactic.append(cheapest_robot)
-            tactic.set_assigned_robots(robots_for_tactic)
 
-        for tactic in self.prioritized_tactics:
-            tactic.init_roles(world_state)
+            if robots_for_tactic is not None:
+                used_robots.update(robots_for_tactic)
+                tactic.set_assigned_robots(robots_for_tactic)
+                self.approved_prioritized_tactics.append(tactic)
+                tactic.init_roles(world_state)
+            else:
+                numRobotsAvailable = (
+                    len(world_state.our_robots)
+                    - len(used_robots)
+                    - numOfInvisibleRobots
+                )
+                print(
+                    f"Tactic denied: {len(tactic.role_requests)} requested roles, but only {numRobotsAvailable} robots available"
```

Comment:
nit: add tactic name to the debug output
```suggestion
                    f"Tactic {tactic} denied: {len(tactic.role_requests)} requested roles, but only {numRobotsAvailable} robots available"
```